chia/full_node/block_store.py

```diff
@@ -134,15 +131,17 @@ def maybe_decompress(self, block_bytes: bytes) -> FullBlock:
 
     async def rollback(self, height: int) -> None:
         if self.db_wrapper.db_version == 2:
-            await self.db.execute(
-                "UPDATE OR FAIL full_blocks SET in_main_chain=0 WHERE height>? AND in_main_chain=1", (height,)
-            )
+            async with self.db_wrapper.write_db() as conn:
+                await conn.execute(
+                    "UPDATE OR FAIL full_blocks SET in_main_chain=0 WHERE height>? AND in_main_chain=1", (height,)
+                )
 
     async def set_in_chain(self, header_hashes: List[Tuple[bytes32]]) -> None:
         if self.db_wrapper.db_version == 2:
-            await self.db.executemany(
-                "UPDATE OR FAIL full_blocks SET in_main_chain=1 WHERE header_hash=?", header_hashes
-            )
+            async with self.db_wrapper.write_db() as conn:
```

Comment:
It's very important that this is called inside the transaction along with set_in_chain, etc etc. Is it OK that we are just re-acquiring the lock here? I would think it's safe to actually pass in the DBWrapper2 instead, and disallow nesting.

Reply:
Yes, pass the Connection object.
What I am worried about is that this API kind of encourages people to call these methods like 'set_in_chain` and `rollback` without grabbing a Connection, which doesn't seem safe. Wouldn't it be better for this method to require Connection as argument?

Reply:
I agree. It would be nice to not have to rely on nested transactions at all. It would probably create some more clarity in the code as well. Maybe `BlockStore` shouldn't even have a db wrapper member at all.

I'm hesitant to pile that onto this PR though, it's already pretty big.

Reply:
I think there is an opportunity to enhance the clarity of proper usage.  I think this does not cause any significant regression in that regard though.  Add it to the follow-up list I guess?  https://github.com/Chia-Network/chia-blockchain/pull/10166#issuecomment-1070867445